src/cudadecoder/batched-threaded-nnet3-cuda-online-pipeline.cc

```diff
@@ -0,0 +1,478 @@
+  feature_info_->ivector_extractor_info.greedy_ivector_extractor = true;
+
+  OnlineNnet2FeaturePipeline feature(*feature_info_);
+  use_ivectors_ = (feature.IvectorFeature() != NULL);
+  input_dim_ = feature.InputFeature()->Dim();
+  if (use_ivectors_) ivector_dim_ = feature.IvectorFeature()->Dim();
+  model_frequency_ = feature_info_->GetSamplingFrequency();
+  BaseFloat frame_shift = feature_info_->FrameShiftInSeconds();
+  input_frames_per_chunk_ = config_.compute_opts.frames_per_chunk;
+  seconds_per_chunk_ = input_frames_per_chunk_ * frame_shift;
+  int32 samp_per_frame = static_cast<int>(model_frequency_ * frame_shift);
+  samples_per_chunk_ = input_frames_per_chunk_ * samp_per_frame;
+  BatchedStaticNnet3Config nnet3_config;
+  nnet3_config.compute_opts = config_.compute_opts;
+  nnet3_config.max_batch_size = max_batch_size_;
+  nnet3_config.nchannels = config_.num_channels;
+  nnet3_config.has_ivector = (feature.IvectorFeature() != NULL);
+
+  cuda_nnet3_.reset(new BatchedStaticNnet3(nnet3_config, *am_nnet_));
+  output_frames_per_chunk_ = cuda_nnet3_->GetNOutputFramesPerChunk();
+}
+
+void BatchedThreadedNnet3CudaOnlinePipeline::FinalizeDecoding(
+    int32 ichannel, CorrelationID corr_id) {
+  Lattice lat;
+  cuda_decoder_->ConcurrentGetRawLatticeSingleChannel(ichannel, &lat);
+
+  // Getting the channel callback now, we're going to free that channel
+  std::unique_ptr<std::function<void(CompactLattice &)>> callback;
+  callback = std::move(channels_callbacks_[ichannel]);
+  // Done with this channel. Making it available again
+  {
+    std::lock_guard<std::mutex> lk(available_channels_m_);
+    available_channels_.push_back(ichannel);
+  }
+
+  // If necessary, determinize the lattice
+  CompactLattice dlat;
+  if (config_.determinize_lattice) {
+    DeterminizeLatticePhonePrunedWrapper(*trans_model_, &lat,
+                                         config_.decoder_opts.lattice_beam,
+                                         &dlat, config_.det_opts);
+  } else {
+    ConvertLattice(lat, &dlat);
+  }
+
+  if (dlat.NumStates() == 0) {
+    KALDI_WARN << "Empty lattice.";
+  } else {
+    if (word_syms_) {
+      CompactLattice best_path_clat;
+      CompactLatticeShortestPath(dlat, &best_path_clat);
+
+      Lattice best_path_lat;
+      ConvertLattice(best_path_clat, &best_path_lat);
+
+      std::vector<int32> alignment;
+      std::vector<int32> words;
+      LatticeWeight weight;
+      GetLinearSymbolSequence(best_path_lat, &alignment, &words, &weight);
+      std::ostringstream oss;
+      for (size_t i = 0; i < words.size(); i++) {
+        std::string s = word_syms_->Find(words[i]);
+        if (s == "") oss << "Word-id " << words[i] << " not in symbol table.";
+        oss << s << " ";
+      }
+      {
+        std::lock_guard<std::mutex> lk(stdout_m_);
+        KALDI_LOG << "OUTPUT: " << oss.str();
+      }
+    }
+
+    // if ptr set and if callback func callable
+    if (callback && *callback) {
+      (*callback)(dlat);
```

Comment:
Would it be better if the callback is always called, even when the lattice is empty? Not knowing whether the callback will fire depending on the contents of the utterance makes programming against it difficult.

Reply:
I see. You're right. I'll take those lines out of the if